protocol/scripts/genesis/prod_pregenesis.sh

```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+set -eo pipefail
+
+# This script creates the pregenesis file for a production network.
+#
+# The script must be run from the root of the `v4-chain` repo.
+#
+# example usage:
+# $ ./scripts/genesis/prod_pregenesis.sh ./build/dydxprotocold
+
+# Check for missing required arguments
+if [ -z "$1" ]; then
+  echo "Error: Missing required argument DYDX_BINARY."
+  echo "Usage: $0 <DYDX_BINARY> [-s|--SEED_FAUCET_USDC]"
+  exit 1
+fi
+
+# Capture the required argument
+DYDX_BINARY="$1"
+
+TMP_CHAIN_DIR="/tmp/prod-chain"
+TMP_EXCHANGE_CONFIG_JSON_DIR="/tmp/prod-exchange_config"
+FIFTEEN_ZEROS="000000000000000"
+EIGHTEEN_ZEROS="000$FIFTEEN_ZEROS"
+BRIDGE_MODACC_BALANCE="1000000000$EIGHTEEN_ZEROS" # 1e27
+BRIDGE_MODACC_ADDR="dydx1zlefkpe3g0vvm9a4h0jf9000lmqutlh9jwjnsv"
+
+source "./testing/genesis.sh"
+
+# TODO(GENESIS): Update below values before running this script. Sample values are shown.
+################## Start of required values to be updated ##################
+CHAIN_ID="dydx-1"
+# Base denomination of the native token. Usually comes with a prefix "u-", "a-" to indicate unit.
+NATIVE_TOKEN="adv4tnt"
+# Denomination of the native token in whole coins.
+NATIVE_TOKEN_WHOLE_COIN="dv4tnt"
+# Market ID in the oracle price list for the rwards token.
+REWARDS_TOKEN_MARKET_ID=11
+# The numerical chain ID of the Ethereum chain for bridge daemon to query.
+ETH_CHAIN_ID=1
+# The address of the Ethereum contract for bridge daemon to monitor for logs.
+ETH_BRIDGE_ADDRESS="0xcca9D5f0a3c58b6f02BD0985fC7F9420EA24C1f0"
+# The next event id (the last processed id plus one) of the logs from the Ethereum contract.
+BRIDGE_GENESIS_ACKNOWLEDGED_NEXT_ID=0
+# The Ethereum block height of the most recently processed bridge event.
+BRIDGE_GENESIS_ACKNOWLEDGED_ETH_BLOCK_HEIGHT=0
+# Genesis time of the chain.
+GENESIS_TIME="2023-12-31T00:00:00Z"
+# Start time of the community vesting schedule.
+COMMUNITY_VEST_START_TIME="2001-01-01T00:00:00Z"
+# End time of the community vesting schedule.
+COMMUNITY_VEST_END_TIME="2050-01-01T00:00:00Z"
+# Start time of the rewards vesting schedule.
+REWARDS_VEST_START_TIME="2001-01-01T00:00:00Z"
+# End time of the rewards vesting schedule.
+REWARDS_VEST_END_TIME="2050-01-01T00:00:00Z"
+
+################## Start of required values to be updated ##################
+
+cleanup_tmp_dir() {
+	if [ -d "$TMP_EXCHANGE_CONFIG_JSON_DIR" ]; then
+		rm -r "$TMP_EXCHANGE_CONFIG_JSON_DIR"
+	fi
+	if [ -d "$TMP_CHAIN_DIR" ]; then
+		rm -r "$TMP_CHAIN_DIR"
+	fi
+}
+
+# Set production default genesis params.
+function overwrite_genesis_production() {	
```

Comment:
I believe this was the original meta point you raised, but I wonder if we can do something better for now.

Can we add overrides for all the modules? For example, `blocktime`. I think it would be generally easier to reason about that this script will override all the module parameters rather than having to refer to the `testing/genesis.sh` or the default param.

Later we can think about refactoring these.

Reply:
IMO we should keep minimal logic in `overwrite_genesis_production`. It should only contain values that *must* live in this method, which includes
- Values under `TODO(GENESIS)` that need to be updated by genesis coordination party
- Values that are different between testnets and a production network

Reason for keeping `overwrite_genesis_production` minimal:
- It's not covered by any test, whereas `DefaultParams` in Golang code are.
- It's impractical to override all modules (at that point we are just manually specifying the whole genesis.json file)
- It's not the responsibility of this script to reason that necessary parameters are overridden. Instead, anyone should only rely on the final genesis file to reason if parameters are set up correctly, since that's the final source of truth.

Reply:
Generally, testnet-specific parameters should be done in this separate [overwrite_genesis_public_genesis](https://github.com/dydxprotocol/v4-chain/blob/1eefd073950dfe937341973d8ef0791360e4945a/protocol/testing/testnet-external/pregenesis.sh#L135) method, so I agree there's concern but the risk is low.

We also have a [ticket](https://linear.app/dydx/issue/CORE-632/workflow-to-ensure-sample-pregenesisjson-in-sync-with-prod-genesissh) to ensure the `sample_pregenesis.json` is in sync with the `prod_pregenesis.sh` script, which will prevent unexpected side effect